pin/gc/gc.go

```diff
@@ -33,146 +27,232 @@ type Result struct {
 
-	ls = ls.GetOfflineLinkService()
+	roots []pin.Source
+}
+
+var _ GC = (*gctype)(nil)
+
+// NewGC creates new instance of garbage collector
+func NewGC(bs bstore.GCBlockstore, ls dag.LinkService) (GC, error) {
+	return &gctype{
+		bs: bs,
+		ls: ls.GetOfflineLinkService(),
+	}, nil
+}
 
+// AddPinSource adds as pin.Source to be considered by the GC.
+// Any calls to AddPinSource have to be done before any calls to Run.
+func (g *gctype) AddPinSource(s ...pin.Source) error {
+	g.roots = append(g.roots, s...)
+	sort.SliceStable(g.roots, func(i, j int) bool {
+		return g.roots[i].SortValue() < g.roots[j].SortValue()
+	})
+
+	return nil
+}
+
+// Run starts the garbage collector and continues it async
+func (g *gctype) Run(ctx context.Context) <-chan Result {
 	output := make(chan Result, 128)
+	tri := newTriset()
 
-	go func() {
-		defer close(output)
-		defer unlocker.Unlock()
-		defer elock.Done()
+	emark := log.EventBegin(ctx, "GC.mark")
 
-		gcs, err := ColoredSet(ctx, pn, ls, bestEffortRoots, output)
-		if err != nil {
-			output <- Result{Error: err}
-			return
+	for _, r := range g.roots {
+		// Stop adding roots at first Direct pin
+		if r.Direct {
+			break
 		}
-		emark.Append(logging.LoggableMap{
-			"blackSetSize": fmt.Sprintf("%d", gcs.Len()),
-		})
-		emark.Done()
-		esweep := log.EventBegin(ctx, "GC.sweep")
-
-		keychan, err := bs.AllKeysChan(ctx)
+		cids, err := r.Get()
 		if err != nil {
 			output <- Result{Error: err}
-			return
+			return output
 		}
-
-		errors := false
-		var removed uint64
-
-	loop:
-		for {
-			select {
-			case k, ok := <-keychan:
-				if !ok {
-					break loop
-				}
-				if !gcs.Has(k) {
-					err := bs.DeleteBlock(k)
-					removed++
-					if err != nil {
-						errors = true
-						output <- Result{Error: &CannotDeleteBlockError{k, err}}
-						//log.Errorf("Error removing key from blockstore: %s", err)
-						// continue as error is non-fatal
-						continue loop
-					}
-					select {
-					case output <- Result{KeyRemoved: k}:
-					case <-ctx.Done():
-						break loop
-					}
-				}
-			case <-ctx.Done():
-				break loop
-			}
+		for _, c := range cids {
+			tri.InsertGray(c, r.Strict)
 		}
-		esweep.Append(logging.LoggableMap{
-			"whiteSetSize": fmt.Sprintf("%d", removed),
-		})
-		esweep.Done()
-		if errors {
-			output <- Result{Error: ErrCannotDeleteSomeBlocks}
-		}
-	}()
+	}
+	go g.gcAsync(ctx, emark, output, tri)
 
 	return output
 }
 
-func Descendants(ctx context.Context, getLinks dag.GetLinks, set *cid.Set, roots []*cid.Cid) error {
-	for _, c := range roots {
-		set.Add(c)
+func (g *gctype) gcAsync(ctx context.Context, emark *logging.EventInProgress,
+	output chan Result, tri *triset) {
 
-		// EnumerateChildren recursively walks the dag and adds the keys to the given set
-		err := dag.EnumerateChildren(ctx, getLinks, c, set.Visit)
-		if err != nil {
-			return err
+	defer close(output)
+
+	bestEffortGetLinks := func(ctx context.Context, cid *cid.Cid) ([]*node.Link, error) {
```

Comment:
Can we combine this with the below `getLinks` function, using a flag to handle the `err != dag.ErrNotFound` case?

Reply:
You are right, two types of GetLinks is cruft from previous (or currently used GC), I should had eliminated it.